Add QgsAbstractGeometry::simplifiedTypeRef

nyalldawson opened the pull request:
Returns a reference to the simplest lossless representation of this geometry,
e.g. if the geometry is a multipart geometry type with a single member geometry,
a reference to that part will be returned.

This method employs the following logic:

- For multipart geometries containing a single part only a direct reference to that part will be returned.
- For compound curve geometries containing a single curve only a direct reference to that curve will be returned.

This method returns a reference only, and does not involve any geometry cloning.

Designed to allow optimisation by avoiding unnecessary cloning/segmentation, as is done in the 3d renderer here.

nyalldawson committed:
Add QgsAbstractGeometry::simplifiedTypeRef

Returns a reference to the simplest lossless representation of this geometry,
e.g. if the geometry is a multipart geometry type with a single member geometry,
a reference to that part will be returned.

This method employs the following logic:

- For multipart geometries containing a single part only a direct reference to that part will be returned.
- For compound curve geometries containing a single curve only a direct reference to that curve will be returned.

This method returns a reference only, and does not involve any geometry cloning.

m-kuhn commented on python/core/auto_generated/geometry/qgsabstractgeometry.sip.in:
```diff
@@ -719,6 +719,22 @@ Converts the geometry to a specified type.
+- For multipart geometries containing a single part only a direct reference to that part will be returned.
+- For compound curve geometries containing a single curve only a direct reference to that curve will be returned.
+
+This method returns a reference only, and does not involve any geometry cloning.
```
Should we add something like the following to be in line with our standard taxonomy.

> The ownership is not transferred.

nyalldawson replied:
Done!

nyalldawson committed:
Dox

nyalldawson committed:
Use simplifiedTypeRef in 2d geometry rendering

This avoids unnecessary cloning/segmentation in some cases, e.g. if
a compound curve containing only a single linestring curve is rendered.
